src/query/api/v1/httpd/handler.go

```diff
@@ -318,19 +310,78 @@ func (h *Handler) RegisterRoutes() error {
 					Tagged(remoteSource).
 					Tagged(experimentalAPIGroup),
 			)
-			h.router.HandleFunc(annotated.WriteURL,
-				wrapped(experimentalAnnotatedWriteHandler).ServeHTTP,
-			).Methods(annotated.WriteHTTPMethod)
+			wrappedRouteFn(annotated.WriteURL, experimentalAnnotatedWriteHandler,
+				annotated.WriteHTTPMethod)
 		}
 	}
 
 	h.registerHealthEndpoints()
 	h.registerProfileEndpoints()
 	h.registerRoutesEndpoint()
 
+	// Register custom endpoints.
+	for _, custom := range h.customHandlers {
+		for _, method := range custom.Methods() {
+			routeName := routeName(custom.Route(), method)
+			route := h.router.Get(routeName)
+			var prevHandler http.Handler
+			if route != nil {
+				prevHandler = route.GetHandler()
+			}
+			customHandler, err := custom.Handler(nativeSourceOpts, prevHandler)
+			if err != nil {
+				return fmt.Errorf("failed to register custom handler with path %s: %w",
+					routeName, err)
+			}
+
+			if route == nil {
+				wrappedRouteFn(custom.Route(), customHandler, method)
+			} else {
+				route.Handler(wrapped(customHandler))
+			}
+		}
+	}
+
 	return nil
 }
 
+func (h *Handler) addRoute(
+	router *mux.Router,
+	path string,
+	handler http.Handler,
+	methods []string,
+) {
+	h.addRouteHandlerFn(router, path, handler.ServeHTTP, methods...)
+}
+
+func (h *Handler) addRouteHandlerFn(
+	router *mux.Router,
+	path string,
+	handlerFn http.HandlerFunc,
+	methods ...string,
+) {
+	for _, method := range methods {
+		routeName := routeName(path, method)
+		if previousRoute := router.Get(routeName); previousRoute != nil {
+			h.logger.Warn("route already exists",
```

Comment:
I've been thinking a bit about this, and ultimately I think it's better if we do error... it'll be annoying to add all the error checks, but the alternative is really bad operationally

e.g. if we accidentally shadow a handler, and are wondering why it's not being applied, all we have is a single log statement at the very start of the node coming up which is easy to miss and non-trivial to someone who doesn't know what to look for. Hard erroring on a duplicate here is much more obvious to debug + fix from the operational side

Reply:
Agree, let's go with the safest approach.